### DISCUSSION TITLE
Making cppad_lib a shared library (DLL) on windows

### DISCUSSION THREAD

**Discussion by bradbell** [coin-or/CppAD#97] [Features]
There are issues with making `cppad_lib` a share library (i.e., DLL) on windows; see the folllowing pull request
https://github.com/coin-or/CppAD/pull/61

This discussion is about fixing the problem.

**Comment by bradbell** [comment]
If someone wants to fix this, they should be able to find all the static setting using the following command:
`git grep CPPAD_ASSERT_FIRST_CALL_NOT_PARALLEL`

**Comment by bradbell** [comment]
It would also help this effort to set up automated testing of the Visual Studio build  ( using appveyor ?) . 
A way to manually run these tests is prescribed under the heading Visual Studio on
https://coin-or.github.io/CppAD/doc/cmake.htm#CMake%20Command.Visual%20Studio

The current appveyor tests on windows are for minsys; see
https://github.com/coin-or/CppAD/blob/master/appveyor.yml
https://github.com/coin-or/CppAD/blob/master/bin/appveyor.sh

**Reply by jcarpent** [reply]
I've made some efforts in this direction a while ago: https://github.com/coin-or/CppAD/pull/56
Hope it would be helpful.

**Reply by bradbell** [reply]
> I've made some efforts in this direction a while ago: #56
> Hope it would be helpful.

I have left the `CPPAD_LIB_EXPORT` usages in the include files and only changed it's definition to empty in define.hpp; see the results of the following command:
`git grep CPPAD_LIB_EXPORT`
and the following section of define.hpp:
```
/*
This commented out code is for building windows shared libraries which
currently does not work for CppAD:
# ifdef  _MSC_VER
# ifdef  cppad_lib_EXPORTS
# define CPPAD_LIB_EXPORT __declspec(dllexport)
# else
# define CPPAD_LIB_EXPORT __declspec(dllimport)
# endif  // cppad_lib_EXPORTS
# else   // _MSC_VER
# define CPPAD_LIB_EXPORT
# endif
*/
# define CPPAD_LIB_EXPORT
```

The only other changes I have made to suppress building windows shared libraries is in cppad_lib/CMakeLists.txt where the following text occurs:
```
IF( ${CMAKE_SYSTEM_NAME} STREQUAL "Windows" )
    MESSAGE( STATUS "Windows system so building static cppad_lib")
    ADD_LIBRARY( cppad_lib STATIC ${source_list} )
ELSE( ${CMAKE_SYSTEM_NAME} STREQUAL "Windows" )
    MESSAGE( STATUS "Not Windows system so building shared cppad_lib")
    ADD_LIBRARY( cppad_lib SHARED ${source_list} )
    SET_TARGET_PROPERTIES( cppad_lib PROPERTIES SOVERSION ${soversion} )
    #
    FIND_LIBRARY(dl_LIBRARY dl)
    IF( dl_LIBRARY )
        TARGET_LINK_LIBRARIES(cppad_lib ${dl_LIBRARY})
    ENDIF( dl_LIBRARY )
ENDIF( ${CMAKE_SYSTEM_NAME} STREQUAL "Windows" )
```

**Reply by Wang-Peng-debug** [reply]
Hey my friend!
Have you got it finished? I would like to know how to build a DLL for CppAD as well,  and please tell me how if you notice this relpy, that would be of much help for me and I will appreciate it so much!
